### PR TITLE
feat: miners for hard triplets

### PR DESCRIPTION
**Pull request by gvondulong**
This PR introduces multiple hard triplet miners.

### CODE REVIEW

File: finetuner/tuner/pytorch/miner.py [hunk 1]
```diff
@@ -1,6 +1,10 @@
+from cProfile import label
+from multiprocessing.sharedctypes import Value
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:2]: Please clean up unneeded imports

File: finetuner/tuner/pytorch/miner.py [hunk 2]
```diff
@@ -59,6 +63,99 @@ def mine(
...
+        :param distances: A tensor matrix of pairwise distance between each two item
+            embeddings
+
+        :return: three 1D tensors, holding the anchor index, positive index and
+            negative index of each triplet, respectively
+        """
+        # distances = torch.Tensor(
+        #     np.array([[0, 184, 222], [184, 0, 45], [222, 45, 0]])
+        # )  # distances[:3, :3]
+        # labels = labels[:3]
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:106]: Please clean up comments

File: finetuner/tuner/pytorch/miner.py [hunk 3]
```diff
@@ -59,6 +63,99 @@ def mine(
         diffs = matches ^ 1
 
         matches.fill_diagonal_(0)
+
+        triplets = matches.unsqueeze(2) * diffs.unsqueeze(1)
+
+        return torch.where(triplets)
+
+
+class TripletHardMiner(TripletMiner):
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py]: Can you add a docstring for the miner, describing what it does and what the availible strategies are

File: tests/unit/tuner/pytorch/test_miner.py [hunk 4]
```diff
@@ -101,6 +102,44 @@ def test_triplet_miner(labels):
     np.testing.assert_equal(neg_ind.numpy(), true_neg_ind)
 
 
+def test_triplet_easy_hard_miner(labels):
```
Comment (tadejsv) [inline tests/unit/tuner/pytorch/test_miner.py:105]: We will need test for all strategies

File: finetuner/tuner/pytorch/miner.py [hunk 5]
```diff
@@ -59,6 +63,99 @@ def mine(
         diffs = matches ^ 1
 
         matches.fill_diagonal_(0)
+
+        triplets = matches.unsqueeze(2) * diffs.unsqueeze(1)
+
+        return torch.where(triplets)
+
+
+class TripletEasyHardMiner(TripletMiner):
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:72]: Please provide a docstring. Also, do you need to subclass `TripletMiner` - it doesn't seem to share any methods with it.

File: finetuner/tuner/pytorch/miner.py [hunk 6]
```diff
@@ -59,6 +63,99 @@ def mine(
...
+    def __init__(self, strategy="hard"):
+        self.strategy = strategy
+
+    def _get_per_row_min(
+        self, dist_mat: torch.Tensor, semihard_tsh: torch.Tensor = None
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:77]:
```suggestion
        self, dist_mat: torch.Tensor, semihard_tsh: Optional[torch.Tensor] = None
```

File: finetuner/tuner/pytorch/miner.py [hunk 7]
```diff
@@ -59,6 +63,99 @@ def mine(
...
+    ):
+        non_zero_mask = dist_mat > 0
+
+        # Prevent wrong neg. samples from being extracted
+        row_max = torch.max(dist_mat, dim=1, keepdim=True)[0]
+        dist_mat += (row_max + 1) * torch.logical_not(non_zero_mask)
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:83]: Why is adding row max needed?

File: finetuner/tuner/pytorch/miner.py [hunk 8]
```diff
@@ -59,6 +63,99 @@ def mine(
...
+        :return: three 1D tensors, holding the anchor index, positive index and
+            negative index of each triplet, respectively
+        """
+        # distances = torch.Tensor(
+        #     np.array([[0, 184, 222], [184, 0, 45], [222, 45, 0]])
+        # )  # distances[:3, :3]
+        # labels = labels[:3]
+        np.random.seed(1234)
+        distances = torch.Tensor(np.round_(np.random.random((6, 6)), 1))
+        distances -= distances * torch.eye(len(distances))
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:109]: This should be cleaned up

**Review by tadejsv — CHANGES_REQUESTED** [review]

**Review by JoanFM — COMMENTED** [review]
Just as a general question, how ks it decided that a case is hard or not?

**Comment by gvondulong** [comment]
> 
In the image below, all negative samples that fall within the red area are hard create a hard triplet. If a negative sample falls within the orange are, it is considered semihard. Green follows the same logic. 

In the code: 
- For hard triplets, we take a negative sample, that is close, and for the same anchor we take all positive samples that are further away from the negative one. 
- For easy triplets, we basically do the same in reverse. 
- For semihard, we take a negative sample and then take one positive sample that is still closer to the anchor, than the negative sample we picked. 
image

**Comment by bwanglzu** [comment]
> For hard triplets, we take a negative sample, that is close, and for the same anchor we take all positive samples that are further away from the negative one.

question : how do we define close? cosine similarity, or it is `margin` involved?

**Review by JoanFM — COMMENTED** [review]
@tadejsv isnt strategy something that we can add to the miner via composition and not inheritance?

File: finetuner/tuner/pytorch/miner.py (hunk 7 above)
Comment (gvondulong) [reply finetuner/tuner/pytorch/miner.py:90]: Because I am getting the minimum per row. Since there are potentially zeros in the rows, I need to overwrite them with a larger value to make sure they are not extracted.

File: finetuner/tuner/pytorch/miner.py [hunk 9]
```diff
@@ -59,11 +62,151 @@ def mine(
         diffs = matches ^ 1
 
         matches.fill_diagonal_(0)
+
         triplets = matches.unsqueeze(2) * diffs.unsqueeze(1)
 
         return torch.where(triplets)
 
 
+class BaseClassEasyHardMiner:
```
Comment (gvondulong) [inline finetuner/tuner/pytorch/miner.py:71]: I think it would be flatter/more pythonic, if this was simply a function, called smth. along the lines of: `apply_easy_hard_strategy`?
Would you agree?

File: finetuner/tuner/pytorch/miner.py [hunk 10]
```diff
@@ -59,11 +62,151 @@ def mine(
...
+        matches, diffs = self.apply_strategy(matches, diffs, distances)
+        triplets = matches.unsqueeze(2) * diffs.unsqueeze(1)
+
+        return torch.where(triplets)
+
+
+class SiameseEasyHardMiner(BaseClassMiner[torch.Tensor], BaseClassEasyHardMiner):
+    def __init__(self, strategy: str = 'hard'):
+        BaseClassEasyHardMiner.__init__(self, strategy)
+
```
Comment (gvondulong) [inline finetuner/tuner/pytorch/miner.py:171]: Will add the docstrings, once we settled on the implementation.

File: finetuner/tuner/pytorch/miner.py (hunk 9 above)
Comment (tadejsv) [reply finetuner/tuner/pytorch/miner.py:71]: Yes, I would do this

**Comment by gvondulong** [comment]
> > For hard triplets, we take a negative sample, that is close, and for the same anchor we take all positive samples that are further away from the negative one.
> 
> question : how do we define close? cosine similarity, or it is `margin` involved?

Distances are given inside the miner, but in principle any distance metric is possible.

**Review by hanxiao — CHANGES_REQUESTED** [review]
why miner has to be so tightly coupled with torch API? this makes porting to Keras/Paddle unnecessarily hard. Can we simply use numpy API to do the same? I'm not seeing anywhere using GPU anyway, so there is no obvious reason to me to use Pytorch.

**Comment by tadejsv** [comment]
@hanxiao GPU is used everywhere the tensors are

**Comment by gvondulong** [comment]
> why miner has to be so tightly coupled with torch API? this makes porting to Keras/Paddle unnecessarily hard. Can we simply use numpy API to do the same? I'm not seeing anywhere using GPU anyway, so there is no obvious reason to me to use Pytorch.

It is true that the GPUs should be used everywhere, since the input tensors will be on GPU during training. 
But I agree that we should think about sacrificing a little performance (hard mining is not really heavy) for reusability. 
Having the implementation in numpy, would alleviate the trouble of implementing it three times.

**Comment by hanxiao** [comment]
in general, i'd like to first see this strategy really works, meaning that it can considerably improve the finetuned performance.

After we are sure about the performance improvement, we can think of generalize it to Keras and Paddle

**Review by JoanFM — CHANGES_REQUESTED** [review]
The structure of the code seems to miss the concept of `pytorch` as `metadata` language and the concept of `pytorch` as the underlying model.

This needs to be clarified and tested how it works for tuning on all frameworks.

**Review by JoanFM — CHANGES_REQUESTED** [review]
`miner` should go under `tuner` and not any specific framework

**Review by JoanFM — COMMENTED** [review]
This is not what was agreed right?

**Comment by gvondulong** [comment]
> This is not what was agreed right?

Can you elaborate?

**Comment by JoanFM** [comment]
> > This is not what was agreed right?
> 
> Can you elaborate?

We can merge now, but as soon as a new mining strategy is impleemnted, we move to only implement `mining` in `torch` as it is our meta-language but not because it is tight to train torch models

**Comment by gvondulong** [comment]
> > > This is not what was agreed right?
> > 
> > 
> > Can you elaborate?
> 
> We can merge now, but as soon as a new mining strategy is impleemnted, we move to only implement `mining` in `torch` as it is our meta-language but not because it is tight to train torch models

Well, the intention behind the refactor, in the last commits, was to separate the mining form the tuners. Now the `StrategicMiningHelper` class sits in `tuner` and is used in each submodule across all frameworks. There is now a wrapper, written in each of the frameworks, encloses the mining procedure (written in torch). Hope this answers the question. :)

File: finetuner/tuner/pytorch/miner.py [hunk 11]
```diff
@@ -59,6 +60,138 @@ def mine(
...
+            negative index of each triplet, respectively
+        """
+        assert len(distances) == len(labels)
+
+        labels1, labels2 = labels.unsqueeze(1), labels.unsqueeze(0)
+        matches = (labels1 == labels2).byte()
+        diffs = matches ^ 1
+
+        matches.fill_diagonal_(0)
+        print("TORCH MATCHES/DIFFS: ", matches, diffs)
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:189]: This should be removed

File: finetuner/tuner/paddle/miner.py [hunk 12]
```diff
@@ -97,6 +194,82 @@ def mine(
...
+        # Apply mining strategy
+        updated_matches, updated_diffs = self.strategic_mining_helper.apply_strategy(
+            torch.Tensor(matches.numpy()),
+            torch.Tensor(diffs.numpy()),
+            torch.Tensor(distances.numpy()),
+            to_numpy=True,
+        )
+
+        matches = paddle.Tensor(updated_matches)
+        diffs = paddle.Tensor(updated_diffs)
```
Comment (tadejsv) [inline finetuner/tuner/paddle/miner.py:260]: You are not properly setting device. Also, same about `Tensor`

File: finetuner/tuner/paddle/miner.py [hunk 13]
```diff
@@ -97,6 +194,82 @@ def mine(
...
+        if labels.size == 0:
+            return _empty_tensor(), _empty_tensor(), _empty_tensor()
+
+        labels1, labels2 = labels.unsqueeze(1), labels.unsqueeze(0)
+        matches = paddle.cast(labels1 == labels2, 'int32')
+        diffs = 1 - matches
+
+        # Apply mining strategy
+        updated_matches, updated_diffs = self.strategic_mining_helper.apply_strategy(
+            torch.Tensor(matches.numpy()),
```
Comment (tadejsv) [inline finetuner/tuner/paddle/miner.py:253]: Using `Tensor` instead of `tensor` will convert everything to `float32` - but this is not the desired type, we want `uint8`. Also, you are not setting the device

File: finetuner/tuner/paddle/miner.py [hunk 14]
```diff
@@ -63,6 +65,101 @@ def mine(
...
+
+        l1, l2 = labels.unsqueeze(1), labels.unsqueeze(0)
+        matches = paddle.cast(l1 == l2, 'int32')
+        diffs = 1 - matches
+        matches = paddle.triu(matches, 1)
+        diffs = paddle.triu(diffs)
+
+        # Apply mining strategy
+        updated_matches, updated_diffs = self.strategic_mining_helper.apply_strategy(
+            torch.Tensor(matches.numpy()),
```
Comment (tadejsv) [inline finetuner/tuner/paddle/miner.py:127]: Same here regarding conversion to tensor

File: finetuner/tuner/paddle/miner.py [hunk 15]
```diff
@@ -63,6 +65,101 @@ def mine(
...
+            torch.Tensor(diffs.numpy()),
+            torch.Tensor(distances.numpy()),
+            to_numpy=True,
+        )
+
+        matches = paddle.Tensor(updated_matches)
```
Comment (tadejsv) [inline finetuner/tuner/paddle/miner.py:133]: Same here regarding device/type

File: finetuner/tuner/miner/mining_strategies.py [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+            if self.pos_strategy != 'all':
+                match_mat, _ = self._update_pos_mat(
+                    match_mat, dist_mat, self.pos_strategy
+                )
+            if self.neg_strategy != 'all':
+                diff_mat, _ = self._update_neg_mat(
+                    diff_mat, dist_mat, self.neg_strategy
+                )
+        if to_numpy:
+            return match_mat.cpu().detach().numpy(), diff_mat.cpu().detach().numpy()
```
Comment (tadejsv) [inline finetuner/tuner/miner/mining_strategies.py:230]: No point in `detach`

File: finetuner/tuner/pytorch/miner.py [hunk 17]
```diff
@@ -59,6 +60,138 @@ def mine(
...
+        matches, diffs = self.strategic_mining_helper.apply_strategy(
+            matches, diffs, distances
+        )
+        print("TORCH UPDATED MATCHES/DIFFS: ", matches, diffs)
```
Comment (tadejsv) [inline finetuner/tuner/pytorch/miner.py:193]: Same here

**Comment by gvondulong** [comment]
@tadejsv on the issue you raised with converting the tensors to float: unfortunately I need to compare the values to `inf` tp make the logic work: So I fear that mining will need to work with float values.

File: finetuner/tuner/keras/miner.py [hunk 18]
```diff
@@ -63,6 +144,82 @@ def mine(
...
+        # Apply mining strategy
+        updated_matches, updated_diffs = self.strategic_mining_helper.apply_strategy(
+            torch.Tensor(matches.numpy()),
+            torch.Tensor(diffs.numpy()),
+            torch.Tensor(distances.numpy()),
+            to_numpy=True,
+        )
+        # Send updated tensors to GPU if available
+        if len(tf.config.list_physical_devices('GPU')) > 0:
+            with tf.device('/CPU:0'):
```
Comment (tadejsv) [inline finetuner/tuner/keras/miner.py:208]: We can remove this - TF device is set globally

File: finetuner/tuner/keras/miner.py [hunk 19]
```diff
@@ -63,6 +144,82 @@ def mine(
...
+        matches = tf.cast(l1 == l2, tf.uint8)
+        diffs = matches ^ 1
+
+        matches = tf.linalg.set_diag(matches, diagonal=tf.zeros_like(labels, tf.uint8))
+
+        # Apply mining strategy
+        updated_matches, updated_diffs = self.strategic_mining_helper.apply_strategy(
+            torch.Tensor(matches.numpy()),
+            torch.Tensor(diffs.numpy()),
+            torch.Tensor(distances.numpy()),
```
Comment (tadejsv) [inline finetuner/tuner/keras/miner.py:203]: Here you need to set device

**Review by tadejsv — COMMENTED** [review]
Last two things, otherwise good to merge

**Review by tadejsv — APPROVED** [review]
LGTM

**Review by JoanFM — APPROVED** [review]
